### ISSUE TITLE
Invoices

### ISSUE THREAD

**Issue by dirceubb** [snipe/snipe-it#6678]
- Snipe-IT Version: v4.6.3
 - OS: Ubuntu Server 18.04.1 LTS
 - Web Server: Apache
 - PHP Version: 7.2.10

**Describe the solution you'd like**
It would be great to be able to create invoices. The invoices should have the following fields:
Invoice Id, provider, date of generation, paid date, total before taxes, taxes, total after taxes, comments and a field to upload the documents
The assets, licenses, components and accesories should be able to be related to an invoice.

**Describe alternatives you've considered**
The workaraund we use right now to fulfill this need is to upload attachments to assets.  The problem is that we have to upload the same attachment to different assets (when many assets belong to one invoice)

**Comment by dirceu3h** [comment]
It is really relevant

**Comment by inventorymadness** [comment]
Also interested in this feature

**Comment by Wosat1** [comment]
Agreed. This is the one feature that Snipe-IT is missing.
The ability to relate one invoice to many assets completes the asset management capability.

**Comment by dirceu3h** [comment]
I think this issue is important and would like you to open it, but if you don't think is important and are not considering to develop it, please leave it closed. Regards

**Comment by jtblsngm** [comment]
I am also interested in this feature.

**Comment by IT-Chaos** [comment]
This is still relevant.

**Comment by FelixPWG** [comment]
Bump! Indeed still relevant @snipe

**Comment by 4cedes** [comment]
Would be a great feature :)

**Comment by Ajtak** [comment]
This is still relevant.

**Comment by ghost** [comment]
I second this. It's very important. Any way to push/support this?

**Comment by snipe** [comment]
It’s not currently on the short term roadmap

**Comment by davilopl** [comment]
Would be a great feature :) I'm interesting too

**Comment by benskigomez** [comment]
Hoping this is still on the short term roadmap. It would be very useful.

**Comment by BOBGEN2** [comment]
This is the only big thing missing from Snipe-IT at the moment, 
We already have the option to click on order number and bring up the other assets, Would be awesome to upload a file there once you have opened the order number.
@snipe Looks like a lot of people want this, Can it be added to the roadmap?

**Comment by Wosat1** [comment]
Just a post to keep this alive. It would be a great feature to have.

**Comment by Kvoo** [comment]
I am also very interested in this.

**Comment by Branislav42** [comment]
Very much the only thing missing coming from GLPI, even a simple "Documents" item that would allow a PDF upload and then link assets to it would be great.

Say you buy 10 pcs in one order, you add the invoice to Snipe IT and link all the 10 pcs to it. Then you can easily track when/where was each of those PCs purchased.

**Comment by Bleauhaus** [comment]
This would be a great feature

**Comment by jodybentley** [comment]
We wanted this feature as well — a single uploaded invoice document shared by every asset purchased on it, rather than duplicating the attachment on each asset. I have an implementation ready and am opening a PR shortly: a first-class Invoice entity (invoice number, supplier, invoice/payment dates, subtotal/tax/total, order number, notes, document uploads via the existing files pipeline), with a nullable invoice_id on assets, an invoice picker on the asset edit form, an assets tab on the invoice view, and full web + API CRUD with tests. Scoped to assets for now — licenses/components/accessories could follow the same pattern if there's interest.

**Comment by snipe** [comment]
Invoices as a first class model sort of doesn't make sense, IMHO. You're not generating invoices for other people, it's just a shared file upload.

**Comment by jodybentley** [comment]
Fair point — "invoice" is probably the wrong frame. The underlying ask in this thread is: one uploaded document (invoice, delivery note, warranty doc) attachable to many assets, instead of uploading duplicates to each. Would you take a PR for that as a shared upload — one file on disk, appearing in each linked asset's Files tab, with a "shared with N assets" view? And if so, would you rather see it as linked upload log entries pointing at one stored file, or is there a direction you'd prefer for the files layer? Happy to build to whichever shape works.

**Comment by Kvoo** [comment]
Spot on jody. Doesn't have to be invoice specific, it would just be nice to have  documents that pertain to multiple assets be shared among them. Our use case is also just to have the invoice/purchase confirmation we received attached to the assets.

**Comment by snipe** [comment]
>  Doesn't have to be invoice specific, it would just be nice to have  documents that pertain to multiple assets be shared among them.

Uploading files to asset models already lets you do that

**Comment by jodybentley** [comment]
True — model files do surface on every asset of that model, and that works when a document applies model-wide (manuals, spec sheets). The gap is documents scoped to a *subset* of a model's assets: an invoice or delivery note covers the 10 units bought on that order, not the other 200 of the same model purchased over the years (or units of different models on the same order). Today the only options are uploading the same file to each of the 10 assets individually, or putting it on the model where it's wrong for the rest.

Would you take a PR that lets one uploaded file be linked to a chosen set of assets — e.g. upload-log entries on each asset pointing at a single stored file, so it just shows in each asset's existing Files tab?

**Comment by snipe** [comment]
Honestly, probably not. What happens when one user decides to delete that file, even though other database records reference it. I'd have to see the UI and db changes to decide tho.

I already have an orders/replenish PR partly done, and I'm not sure if that would overlap with this.

**Comment by jodybentley** [comment]
The deletion problem is the crux, agreed. The shape I had in mind: no single asset owns the file — each linked asset gets its own upload-log row pointing at one stored file. "Delete" from an asset's Files tab removes only that asset's link; the file on disk is only removed when the last link goes away. The UI would show "shared with N assets" on the file row, so it's obvious you're unlinking, not destroying. Happy to put together the actual migration and UI screenshots if that would help you decide.

That said — orders/replenish sounds like it might cover this thread's use case directly: if an order links multiple assets and can carry file uploads (the invoice / purchase confirmation), that's exactly what people here are asking for. I'd rather contribute to that than build something parallel. Is that branch visible anywhere?

**Comment by snipe** [comment]
> each linked asset gets its own upload-log row pointing at one stored file

This has been discussed in several other threads. It's one more join on an already incredibly join-heavy system, which we generally try to stay away from - and if single uploads are still handled the way they are now, it creates a pretty big divergence in how uploads in general are handled within the system from one scenario to another.

> Is that branch visible anywhere?

It's not yet - still have some architectural decisions to make before I can push it up.

**Comment by jodybentley** [comment]
Fair enough — diverging how uploads work for one scenario isn't worth it. Sounds like orders/replenish is the natural home for this use case; happy to test or pitch in once it's up. Thanks for talking it through.

**Comment by Kvoo** [comment]
Imho this is still a feature that even the most basic inventory system should have… having assets linked to their corresponding purchase/delivery note. Weird that you seem so resistant to this feature.

**Comment by snipe** [comment]
@Kvoo We are [not an inventory system, we are an asset management system](https://docs.inventree.org/en/stable/concepts/terminology/), and I am not resistant (I literally just said I have a PR in the works) - it's just very messy and there are a lot of moving parts and messing any of it up will cause real pain points for users. Doing this a naive way will cause huge headaches and data-migrations down the line if we're not careful.
